src/bindings/python/flux/future.py

```diff
@@ -15,7 +15,7 @@
 from flux.core.inner import ffi, lib, raw
 
 
-_THEN_HANDLES = set()
+_THEN_HANDLES = {}
```

Comment:
Why the switch from set to dict here?

Reply:
To support "reference counting" (or at least a very hacky form of it).  Initialize the count to `1` when a callback is first set on the future.  Whenever the `then` callback is run, decrement the counter for that future, and whenever `reset` is called on a future, increment the counter.  If the counter hits `0`, delete the reference to the future from the dictionary.    It's not the best abstraction, since a user accidentally calling `reset` multiple times, back-to-back without an intervening future fulfilment will artificially inflate the count and therefore the reference will never be delete.

I'm definitely willing to refactor this if you have any suggestions.

Reply:
That makes a great deal of sense.  I probably would have held the count in the object and removed it from the set on zero or less, but having it this way will work better if we need to support multiple instances at some point.  No reason to mess with it since you have it working.

Reply:
If you are willing, I'd suggest a comment at the definition site to the effect of what you've said above, just to help future code readers...

Reply:
> If you are willing, I'd suggest a comment at the definition site to the effect of what you've said above, just to help future code readers...

Will do!  I'll also update the commit message body.  Sorry for not having that initially.

Which reminds me, I need to tag the streaming RPC issue in the OP.